### DISCUSSION TITLE
✨ EssentialsX Discord: public preview!

### DISCUSSION THREAD

**Discussion by mdcfe** [EssentialsX/Essentials#4084] [General]
After months of hard work by @JRoy ~~and also months of waiting for Discord to finish their API~~, the new EssentialsX Discord module is ready for its first public preview!

EssentialsX Discord is a lightweight bridge between Minecraft servers and your Discord community. You can bridge together chat channels, allowing players to chat between your Minecraft server and Discord channels. You can also privately message online players from Discord, and server admins can monitor the server console remotely and even run console commands from Discord.

EssentialsX Discord also has an API to allow developers to integrate their own chat plugins and hook custom channels (note that this may be subject to breaking changes until 2.19 is released).

## Getting started

**Alpha 4 (`module/discord`, a253c95e78ebeb4930978b972ce174246b7ab073) can be downloaded [here](http://cdn.essentialsx.net/file/essx-cdn/previews/EssentialsX-Discord-alpha-4.zip).** You can see the setup guide [here](https://github.com/EssentialsX/Essentials/blob/module/discord/EssentialsDiscord/README.md).

Note that the setup process is a little different to typical bot setup - EssentialsX Discord requires the commands scope, unlike most bots, so you'll need to use the link provided in the guide above to add your bot to your server.

## Feedback

Got any feedback? Feel free to leave it in the replies below, or in the `#essx-discord-feedback` channel on [MOSS](https://discord.gg/h8CnPSw) ([click here to jump to the channel](https://discord.com/channels/390942438061113344/817518771233095741) if you're already in MOSS).

**Comment by Andre601** [comment]
Something that should be changed:
When using the `show-avatar` setting to use webhooks does the Discord module still attempt to escape markdown formatting like underscores, which results in cases like the one shown below.

Another thing I suggest is to maybe look for a different way to use emojis, because the `:name:` option is not guaranteed to work consistently due to it being a OS-dependand thing to have.
Maybe a placeholder pattern where you define the Unicode that is then translated to the Unicode format (`\u<code>`)?

The image:
![image](https://user-images.githubusercontent.com/11576465/114872299-77aad000-9dfa-11eb-82f0-77887b30bd5d.png)

**Comment by mdcfe** [comment]
**Alpha 2 is now available!** Follow the link in the first post to download this version.

## Changelog

### Additions
  * Added the plugin name to console messages
  * Added config option (show-avatar) to show the players skin on chat/join/leave messages in discord
  * Added config option (show-name) to show the players name on chat/join/leave messages in discord
  * Added minecraft /reply support from messages received in discord via /msg
    * Replies will be DM'd to discord members who have direct messages enabled
  * Added the (hopefully finalized) API, allowing for discord slash command registration (via a service) and sending messages to channel definitions (via the DiscordMessageEvent)
    * See https://git.io/JOnhD for current API. Documentation will be created soon
  * Added PlaceholderAPI support for mc-to-discord, join, quit, and death message formats
  * Added config option (chat.discord-filter) to filter discord messages that match the provided pattern
    * The default provided in the config.yml will filter messages that start with ! and ?
    * To add this behavior to an old config, add this to your config under the chat section: `discord-filter: "^[!?]"`

### Performance
  * Reduced allocations in the console injector
  * Reduced amount of busy waiting for webhook loading

### Bug Fixes
  * Fixed Prefixes and Suffixes not being stripped of color codes in discord
  * Fixed the console relay breaking after using "/ess reload"
  * Fixed user/display names not being sanitized for markdown
  * Fixed incompatibility with Spigot 
  * Tastes less like pineapple (😢)

You can also see a list of commits between alpha 1 and alpha 2 here (includes upstream EssentialsX changes): 859c14c...e2aafb158adc3060c34f83974bd93764440cf073

**Comment by mdcfe** [comment]
~~**Alpha 3 is now available**~~ Edit: **Alpha 4 is now available**, which fixes an issue where channels weren't resolved to the definitions in the config. Download it from [here](http://cdn.essentialsx.net/file/essx-cdn/previews/EssentialsX-Discord-alpha-4.zip).

## Changelog
### Bug fixes
* Fixed channels not being resolved to the correct definition at all
* Allow raw channel IDs in the config file
